### PR TITLE
[core] Ensure failed to register worker is killed and print better log

### LINKED ISSUE #20769: [Feature] Better log when worker not registered
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Logs for worker starts/failed are not very clear which makes it not so easy to debug. We should consider adding when worker start failed:

* Whether the worker is hanging
* Whether the worker is kill/dead
* Add possible reasons for this
* Better logs when workers start

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Before this PR, then raylet notices there is something wrong with the worker starting, it'll start a new worker but not kill the old one. If the old one is hanging, it'll lead to resource waste.
This PR killed the failed worker if it's still alive and also print useful logs

## Related issue number
Closes #20769 #20771

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/util/process.h [hunk 1]
```diff
@@ -80,6 +80,7 @@ class Process {
                               const ProcessEnvironment &env = {});
   static Process CreateNewDummy();
   static Process FromPid(pid_t pid);
+  bool IsDead() const;
```
Comment (jjyao) [inline src/ray/util/process.h:83]: This is not implemented.

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -434,16 +434,25 @@ void WorkerPool::MonitorStartingWorkerProcess(const Process &proc,
...
     auto &state = this->GetStateForLanguage(language);
     // Since this process times out to start, remove it from starting_worker_processes
     // to avoid the zombie worker.
     auto it = state.starting_worker_processes.find(proc_startup_token);
     if (it != state.starting_worker_processes.end()) {
-      RAY_LOG(INFO) << "Some workers of the worker process(" << proc.GetId()
-                    << ") have not registered to raylet within timeout.";
+      RAY_LOG(ERROR)
+          << "Some workers of the worker process(" << proc.GetId()
+          << ") have not registered to raylet within timeout. "
```
Comment (ericl) [inline src/ray/raylet/worker_pool.cc:447]:
```suggestion
          << ") have not registered within the timeout. "
```

File: python/ray/tests/test_failure_4.py [hunk 3]
```diff
@@ -433,5 +434,39 @@ def ready(self):
...
+    # This test is to make sure
+    #   1. when worker failed to register, raylet will print useful log
+    #   2. raylet will kill hanging worker
+    with monkeypatch.context() as m:
+        # delay internal kv for 5s
+        # this delay will make worker start slow
+        m.setenv(
+            "RAY_testing_asio_delay_us",
+            "InternalKVGcsService.grpc_server.InternalKVGet"
+            "=3000000:3000000")
```
Comment (ericl) [inline python/ray/tests/test_failure_4.py:447]: Change this to something higher like 999999999 to make it clear?

**Review by ericl — APPROVED** [review]

File: python/ray/tests/test_failure_4.py (hunk 3 above)
Comment (iycheng) [reply python/ray/tests/test_failure_4.py]: @ericl I'm reverting this back. It looks like internal kv is not just used in worker start, but also driver starts. So make it 9999999... will make it timeout. I think here we need better control for this, but for now, let me just make it 2 seconds.

**Pull request merged**